spec/field-aligned-binary.md

```diff
@@ -0,0 +1,214 @@
+  unsigned, 32-bit integer `i`. `i` must be less than the number of
+  `Alignment` options. It then consumes the `Alignment` at index `i` in the
+  segment.
+
+## Field representation
+
+An alignment-annotated `value: ValueAtom`, valid for alignment `align:
+AlignmentAtom` has a representation as a sequence of field elements:
+
+* If `align = compress`, it is represented as a single field element by
+  applying cryptographically hashing `value`. (length: 1)
+* If `align = field`, it is represented as a single field element by
+  interpreting `value` as a little-endian bigint, modolo the field order. (length: 1)
+* If `align = bytes<n>`, `value` is split into chunks according to the largest
+  number of bytes representable in a single field `k` (currently: 31), each of
+  which are encoded independently as a single field element is in `align =
+  field`. Chunks are filled *from the end* of the byte array: A `value` of
+  length 40, with 31 bytes representable, is represented as two field elements,
+  the first encoding the first 9 bytes, the second the remaining 31. (length: `ceil(n / k)`)
+
+An alignment-annotated `value: Value`, valid for alignment `align: Alignment`
+has a representation as a sequence of field elements, arrived at by consuming
+the alignment, and appending the generated parts in sequence:
+
+* Consuming an `AlignmentAtom` produces the representation above 
+* Consuming an "option"-type `AlignmentSegment` proceeds as consuming one in
+  [Validty](#validity), consuming first the `bytes<4>` representation, then
+  the chosen `Alignment`. The length of each `Alignment` option's field
+  representation is calculated, and if the chosen one is shorter than this, it
+  is padded to this length with zero elements.
+
+A `AlignedValue` can be completely encoded into field elements as follows:
+* Encoding the size length of the `AlignedValue` as a single field element.
+* Encoding each `AlignmentAtom` as follows:
+  * `bytes<n>` encoded as `<n>`
+  * `compress` encoded as `-1`
+  * `field` encoded as `-2`
+* Encoding each `ValueAtom`s with respect to their `AlignmentAtom`.
+
+## High-level vs low-level JS values
+
+Field-Aligned Binary does not contain sufficient structural information to
+output high-level JS values, such as objects, but with type information the
+Abcird compiler should be able to perform automatic mappings between some
+typescript types and the low-level JS values listed above. 
+
+The high(er)-level typescript types understood are:
+
+* Plain objects, e.g. `{foo: "foo", bar: "bar"}`
+* Arrays, e.g. `["foo", "bar"]`
+* Unsigned integer `number`s
+* `true` and `false`
+* Tagged union types, e.g. `{ tag: "foo", bar: number, baz: string } | { tag: "bar" }`
+* `string` and `Uint8Array`
+* Arbitrary types with `<type>.decode(from: Value)` and `<value>.encode(): Value` methods.
+
+Encoding is straight-forward:
+
+* Plain objects are encoded by encoding their fields in sequence.
```

Comment:
Not sure if I mentioned this somewhere before.
The sequence needs to be explicitly defined. In JS the default sequence is by insertion order, which is not sufficient for this spec:
```
Welcome to Node.js v20.18.0.
Type ".help" for more information.
> const obj = {c: "c"}
undefined
> obj.b = 1
1
> obj.a = false
false
> Object.keys(obj)
[ 'c', 'b', 'a' ]
> const obj2 = { c: "c", b: 1, a: false};
undefined
> Object.keys(obj2)
[ 'c', 'b', 'a' ]
```